datafusion/sql/src/select.rs

```diff
@@ -431,6 +432,37 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
         group_by_exprs: Vec<Expr>,
         aggr_exprs: Vec<Expr>,
     ) -> Result<(LogicalPlan, Vec<Expr>, Option<Expr>)> {
+        let schema = input.schema();
```

Comment:
Could you explain why the dependency logic needs to be in the SQL planner? It seems more general than just SQL (it should apply to dataframes as well)

Also, this function calls `LogicalPlanBuilder::aggregate` which then eventally calls `Aggregate::try_new_with_schema` which also re-calculates functional dependency information

Maybe this logic could be put in `Aggregate::try_new_with_schema` if it is not redundant

Reply:
In here, we update group by exprs by their expressions that are in target indices (if they are in the select exprs). With this change, we can use depedents in the functional dependence in the select expression, when their determinant is inside group by expression. We can do so inside `Aggregate::try_new_with_schema` if we pass `select_expr` argument to it. I thought it is a bit weird for aggregate to receive `select_expr` argument. After this PR merges, I can file another PR, that implements this change. Then we can decide which implementation is better. In any case, I have moved groupby update logic to its own function to not clutter this method.

Reply:
> I thought it is a bit weird for aggregate to receive select_expr argument.

Yes I agree that sounds weird. I still don't  understand why the functional dependency can't be computed from the plan itself (e.g. from the annotations on DFSchema and then the structure of the LogicalPlan::Aggregate / LogicalPlan::Project, etc) 

If the logic is in the sql planner, that implies the calculation relies on information that is not encoded in the plan itself (and thus may be lost in subsequent optimizer passes, for example)

> After this PR merges, I can file another PR, that implements this change. Then we can decide which implementation is better. In any case, I have moved groupby update logic to its own function to not clutter this method.

Sounds good

Reply:
> Yes I agree that sounds weird. I still don't understand why the functional dependency can't be computed from the plan itself 

Currently, if the query is
```sql
SELECT sn, amount
            FROM sales_global
            GROUP BY sn
```
and we know that `sn` is primary key.  We rewrite query above as
```sql
SELECT sn, amount
            FROM sales_global
            GROUP BY sn, amount
```
so that select either contains group by expressions or aggregate expressions. This update is done here.
However, if the query were
```sql
SELECT sn
            FROM sales_global
            GROUP BY sn
```
we wouldn't rewrite query as 
```sql
SELECT sn
            FROM sales_global
            GROUP BY sn, amount
```
hence, amount is not used and. We do not need to update group by here. 

For this reason, we need the information of `select exprs` to decide how to extend group by expressions to support a query.
However, if we were to extend group by expressions with their functional dependencies in all cases. We won't need the `select exprs` for this decision. Such as, we would treat 
```sql
SELECT sn
            FROM sales_global
            GROUP BY sn
```
as
```sql
SELECT sn
            FROM sales_global
            GROUP BY sn, amount
```
under the hood.

What do you think about extending group by expression in all cases, without checking for `select exprs`. I think this is a bit confusing.

In short, functional dependency can be calculated without external information. However, query rewrite cannot be done without resorting to the information in the `select exprs` in current design.

Reply:
> What do you think about extending group by expression in all cases, without checking for select exprs. I think this is a bit confusing.

I agree

> In short, functional dependency can be calculated without external information. However, query rewrite cannot be done without resorting to the information in the select exprs in current design.

I think this is the key point I was missing -- that the code in the sql planner is actually doing a SQL level rewrite. Thank you for clarifying. 

What would you think about pulling the rewrite code into its own function and adding a clear explanation of what it was doing (basically copy/paste your example from above)?

Reply:
Thanks @alamb. I have opened a mini [PR](https://github.com/apache/arrow-datafusion/pull/7111) for the update of docstring.